File: Combinatorics/series_verify.py
"""
验证两个无穷级数恒等式：
    268 = Σ [C(2n,n)·H_{n+1}^{(2)} / 2^{3n-1}(n+1)] × [ ... ]
    268 = Σ [H_{n+1}^{(4)} / 2^{5n+3}] × [ ... ]
"""
from decimal import Decimal, getcontext
from math import comb
from functools import lru_cache

getcontext().prec = 200  # 200位十进制精度

# ========== k-step Fibonacci ==========
@lru_cache(maxsize=None)
def F_k(n: int, k: int) -> int:
    """F_m^(k) = 0 (m≤0), F_1^(k) = 1, F_n^(k) = Σ_{j=1}^k F_{n-j}^(k)"""
    if n <= 0:
        return 0
    if n == 1:
        return 1
    return sum(F_k(n - j, k) for j in range(1, k + 1))

# ========== 广义调和数 H_n^(r) ==========
def H(n: int, r: int) -> Decimal:
    """H_n^(r) = Σ_{j=1}^n 1/j^r"""
    if n <= 0:
        return Decimal('0')
    return sum(Decimal(1) / Decimal(j) ** r for j in range(1, n + 1))

# ========== 公式 1 ==========
def formula1(N: int) -> Decimal:
    """验证第一个级数收敛到 268"""
    total = Decimal('0')
    print(f"{'n':>4} | {'a_n':>28} | {'部分和':>30}")
    print("-" * 75)
    for n in range(1, N + 1):
        # 二项式系数（精确整数）
        C = comb(2 * n, n)
        # 广义调和数
        h = H(n + 1, 2)
        # 括号内三项（全部用精确整数运算）
        t1 = 16 * (n + 1) ** 3 * F_k(n + 5, 5)
        t2 = 4 * (2 * n + 1) * (2 * n ** 2 + 6 * n + 5) * F_k(n + 6, 5)
        t3 = (n + 2) * (2 * n + 1) * (2 * n + 3) * F_k(n + 7, 5)
        bracket = t1 - t2 + t3

        # 分母
        denom = Decimal(2) ** (3 * n - 1) * Decimal(n + 1)

        # 单项 = C × h × bracket / denom
        term = Decimal(C) * h * Decimal(bracket) / denom
        total += term

        print(f"{n:>4} | {term:>28.10f} | {total:>30.15f}")
    return total

# ========== 公式 2 ==========
def formula2(N: int) -> Decimal:
    """验证第二个级数收敛到 268"""
    total = Decimal('0')
    print(f"{'n':>4} | {'b_n':>28} | {'部分和':>30}")
    print("-" * 75)
    for n in range(1, N + 1):
        h = H(n + 1, 4)

        # 括号内三项（精确整数）
        t1 = 1024 * (n + 1) ** 4 * comb(4 * n, 2 * n) * F_k(n, 3)
        t2 = 32 * ((n + 1) ** 4 + (n + 2) ** 4) * comb(4 * n + 4, 2 * n + 2) * F_k(n + 1, 3)
        t3 = (n + 2) ** 4 * comb(4 * n + 8, 2 * n + 4) * F_k(n + 2, 3)
        bracket = t1 - t2 + t3

        # 分母
        denom = Decimal(2) ** (5 * n + 3)

        term = h * Decimal(bracket) / denom
        total += term

        print(f"{n:>4} | {term:>28.10f} | {total:>30.15f}")
    return total

if __name__ == "__main__":
    print("=" * 75)
    print("Formula 1: 验证 268 = Σ ... F^(5) ...")
    print("=" * 75)
    r1 = formula1(80)
    print(f"\n→ 前80项部分和 = {r1:.20f}\n")

    print("=" * 75)
    print("Formula 2: 验证 268 = Σ ... F^(3) ...")
    print("=" * 75)
    r2 = formula2(80)
    print(f"\n→ 前80项部分和 = {r2:.20f}\n")

File: Combinatorics/series_verify.cpp
/*
 * 验证两个无穷级数恒等式（GMP 高精度版）
 * 编译: g++ verify.cpp -lgmpxx -lgmp -o verify
 * 运行: ./verify
 */
#include <iostream>
#include <iomanip>
#include <vector>
#include <gmpxx.h>

using namespace std;

// ========== k-step Fibonacci（记忆化） ==========
class KFib {
    int k;
    vector<mpz_class> cache;
public:
    KFib(int k_) : k(k_) { cache.push_back(0); } // cache[0]=0

    mpz_class get(int n) {
        if (n <= 0) return 0;
        if (n == 1) return 1;
        while ((int)cache.size() <= n) {
            int sz = cache.size();
            mpz_class sum = 0;
            for (int j = 1; j <= k && sz - j >= 0; j++)
                sum += cache[sz - j];
            cache.push_back(sum);
        }
        return cache[n];
    }
};

// 二项式系数 C(m,j)
inline mpz_class C(unsigned long m, unsigned long j) {
    mpz_class res;
    mpz_bin_uiui(res.get_mpz_t(), m, j);
    return res;
}

// 广义调和数 H_n^(r)
mpf_class harmonic(int n, int r) {
    if (n <= 0) return mpf_class(0);
    mpf_class sum = 0;
    for (int j = 1; j <= n; j++) {
        mpf_class term = 1;
        for (int p = 0; p < r; p++) term /= mpf_class(j);
        sum += term;
    }
    return sum;
}

int main() {
    mpf_set_default_prec(1024);  // ~300位十进制精度
    mpf_class target(268);

    cout << fixed << setprecision(15);

    // ====== 公式 1 ======
    cout << "Formula 1:\n";
    cout << string(60, '-') << "\n";
    KFib F5(5);
    mpf_class total1 = 0;

    for (int n = 1; n <= 60; n++) {
        mpz_class comb2n = C(2 * n, n);
        mpf_class h = harmonic(n + 1, 2);

        mpz_class t1 = 16; t1 *= (n + 1); t1 *= (n + 1); t1 *= (n + 1); t1 *= F5.get(n + 5);
        mpz_class t2 = 4;  t2 *= (2 * n + 1); t2 *= (2 * n * n + 6 * n + 5); t2 *= F5.get(n + 6);
        mpz_class t3 = 1;  t3 *= (n + 2); t3 *= (2 * n + 1); t3 *= (2 * n + 3); t3 *= F5.get(n + 7);
        mpz_class bracket = t1 - t2 + t3;

        mpz_class denom; mpz_ui_pow_ui(denom.get_mpz_t(), 2, 3 * n - 1); denom *= (n + 1);

        mpf_class term = mpf_class(comb2n) * h * mpf_class(bracket) / mpf_class(denom);
        total1 += term;

        cout << "n=" << setw(2) << n << "  partial = " << total1 << "\n";
    }
    cout << "→ target = 268,  result = " << total1 << "\n\n";

    // ====== 公式 2 ======
    cout << "Formula 2:\n";
    cout << string(60, '-') << "\n";
    KFib F3(3);
    mpf_class total2 = 0;

    for (int n = 1; n <= 40; n++) {
        mpf_class h = harmonic(n + 1, 4);

        mpz_class t1 = 1024;
        for (int i = 0; i < 4; i++) t1 *= (n + 1);
        t1 *= C(4 * n, 2 * n);
        t1 *= F3.get(n);

        mpz_class t2 = 32;
        mpz_class p1 = 1, p2 = 1;
        for (int i = 0; i < 4; i++) { p1 *= (n + 1); p2 *= (n + 2); }
        t2 *= (p1 + p2);
        t2 *= C(4 * n + 4, 2 * n + 2);
        t2 *= F3.get(n + 1);

        mpz_class t3 = 1;
        for (int i = 0; i < 4; i++) t3 *= (n + 2);
        t3 *= C(4 * n + 8, 2 * n + 4);
        t3 *= F3.get(n + 2);

        mpz_class bracket = t1 - t2 + t3;

        mpz_class denom; mpz_ui_pow_ui(denom.get_mpz_t(), 2, 5 * n + 3);

        mpf_class term = h * mpf_class(bracket) / mpf_class(denom);
        total2 += term;

        cout << "n=" << setw(2) << n << "  partial = " << total2 << "\n";
    }
    cout << "→ target = 268,  result = " << total2 << "\n";

    return 0;
}